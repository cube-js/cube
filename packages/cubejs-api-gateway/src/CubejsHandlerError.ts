export class CubejsHandlerError extends Error {
  public constructor(
    public readonly status: number,
    public readonly type: string,
    message: string
  ) {
    super(message || type);
  }
}
