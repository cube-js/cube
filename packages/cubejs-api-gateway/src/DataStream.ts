import * as stream from 'stream';

/**
 * Data stream.
 */
export class DataStream extends stream.Transform {
  public _transform(chunk, encoding, done) {
    this.push(chunk);
    done();
  }

  public _destroy(error: Error | null, callback: (error: Error | null) => void) {
    this.emit('end');
    super._destroy(error, callback);
  }
}

/**
 * Stream response.
 */
export type StreamResponse = {
  types: {
    name: string,
    type: string
  }[],
  rowStream: DataStream,
  release: () => Promise<DataStream>,
};
