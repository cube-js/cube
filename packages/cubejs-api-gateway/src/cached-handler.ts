import type { Handler, Response } from 'express';

type CachedRouterOptions = {
  lifetime: number,
};

interface CachedResponse {
  status: number,
  json: any,
}

export function pipeFromCache(cache: CachedResponse, res: Response) {
  res.status(cache.status)
    .json(cache.json);
}

export function cachedHandler(handler: Handler, options: CachedRouterOptions = { lifetime: 1000 }): Handler {
  let lastCache: CachedResponse = {
    status: 200,
    json: null,
  };
  let lastCacheExpr = new Date(Date.now() - options.lifetime);
  let lock = false;

  const queue: Response[] = [];

  return async (req, res, next) => {
    if (lock) {
      queue.push(res);
    } else {
      if (lastCacheExpr.getTime() > new Date().getTime()) {
        pipeFromCache(lastCache, res);

        return;
      }

      lock = true;

      try {
        const responseWrapper: any = {
          ...res,
          status(code: number) {
            res.status(code);

            lastCache.status = code;

            return responseWrapper;
          },
          json(json: any) {
            res.json(json);

            lastCache.json = json;

            return responseWrapper;
          }
        };

        await handler(
          req,
          responseWrapper,
          next
        );

        lastCacheExpr = new Date(Date.now() + options.lifetime);
        lock = false;
      } catch (e) {
        // console.log('cached-router exception', e);

        lock = false;
        lastCache = {
          status: 200,
          json: null,
        };
        lastCacheExpr = new Date(Date.now() - options.lifetime);

        next(e);
      }

      let queuedResponse: Response | undefined;

      // eslint-disable-next-line no-cond-assign
      while (queuedResponse = queue.pop()) {
        pipeFromCache(lastCache, queuedResponse);
      }
    }
  };
}
