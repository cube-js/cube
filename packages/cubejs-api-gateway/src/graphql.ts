import R from 'ramda';
import moment from 'moment-timezone';

import {
  GraphQLResolveInfo,
  ArgumentNode,
  DirectiveNode,
  FieldNode,
  VariableNode,
  ValueNode,
} from 'graphql';

import {
  objectType,
  enumType,
  extendType,
  list,
  inputObjectType,
  nonNull,
  arg,
  intArg,
  stringArg,
  booleanArg,
  makeSchema as nexusMakeSchema,
  asNexusMethod
} from 'nexus';

import {
  DateTimeResolver,
} from 'graphql-scalars';

import { QUERY_TYPE } from './query';

const DateTimeScalar = asNexusMethod(DateTimeResolver, 'date');

const FloatFilter = inputObjectType({
  name: 'FloatFilter',
  definition(t) {
    t.float('equals');
    t.float('notEquals');
    t.list.float('in');
    t.list.float('notIn');
    t.boolean('set');
    t.float('gt');
    t.float('lt');
    t.float('gte');
    t.float('lte');
  }
});

const StringFilter = inputObjectType({
  name: 'StringFilter',
  definition(t) {
    t.string('equals');
    t.string('notEquals');
    t.list.string('in');
    t.list.string('notIn');
    t.string('contains');
    t.string('notContains');
    t.boolean('set');
  }
});

const DateTimeFilter = inputObjectType({
  name: 'DateTimeFilter',
  definition(t) {
    t.string('equals');
    t.string('notEquals');
    t.list.string('in');
    t.list.string('notIn');
    t.list.string('inDateRange');
    t.list.string('notInDateRange');
    t.string('beforeDate');
    t.string('afterDate');
    t.boolean('set');
  }
});

const OrderBy = enumType({
  name: 'OrderBy',
  members: [
    'asc',
    'desc'
  ]
});

export const TimeDimension = objectType({
  name: 'TimeDimension',
  definition(t) {
    t.nonNull.field('value', {
      type: 'DateTime',
    });
    t.nonNull.field('second', {
      type: 'DateTime',
    });
    t.nonNull.field('minute', {
      type: 'DateTime',
    });
    t.nonNull.field('hour', {
      type: 'DateTime',
    });
    t.nonNull.field('day', {
      type: 'DateTime',
    });
    t.nonNull.field('week', {
      type: 'DateTime',
    });
    t.nonNull.field('month', {
      type: 'DateTime',
    });
    t.nonNull.field('quarter', {
      type: 'DateTime'
    });
    t.nonNull.field('year', {
      type: 'DateTime',
    });
  },
});

function mapType(type: string, isInputType?: boolean) {
  switch (type) {
    case 'time':
      return isInputType ? 'DateTime' : 'TimeDimension';
    case 'string':
      return 'String';
    case 'number':
      return 'Float';
    default:
      return 'String';
  }
}

function mapWhereOperator(operator: string, value: any) {
  switch (operator) {
    case 'in':
      return 'equals';
    case 'notIn':
      return 'notEquals';
    case 'set':
      return (value === true) ? 'set' : 'notSet';
    default:
      return operator;
  }
}

function mapWhereValue(operator: string, value: any) {
  switch (operator) {
    case 'set':
      return undefined;
    default:
      return Array.isArray(value) ? value.map(v => `${v}`) : [`${value}`];
  }
}

function safeName(name: string) {
  return name.split('.').slice(1).join('');
}

function capitalize(name: string) {
  return `${name[0].toUpperCase()}${name.slice(1)}`;
}

function unCapitalize(name: string) {
  return `${name[0].toLowerCase()}${name.slice(1)}`;
}

function applyDirectives(
  directives: readonly DirectiveNode[] | undefined,
  values: Record<string, any>
) {
  if (directives === undefined || directives.length === 0) {
    return true;
  }

  return directives.reduce((result, directive) => {
    directive.arguments?.forEach((argument: ArgumentNode) => {
      if (argument.name.value === 'if') {
        if (
          (directive.name.value === 'include' &&
            !values[(argument.value as VariableNode).name.value]) ||
          (directive.name.value === 'skip' &&
            values[(argument.value as VariableNode).name.value])
        ) {
          result = false;
        }
      }
    });
    return result;
  }, true);
}

function getFieldNodeChildren(node: FieldNode, infos: GraphQLResolveInfo) {
  return (node.selectionSet?.selections.filter((childNode) => (
    childNode.kind === 'Field' &&
    childNode.name.value !== '__typename' &&
    applyDirectives(childNode.directives, infos.variableValues)
  )) || []) as FieldNode[];
}

function parseArgumentValue(value: ValueNode) {
  switch (value.kind) {
    case 'BooleanValue':
    case 'IntValue':
    case 'StringValue':
    case 'FloatValue':
    case 'EnumValue':
      return value.value;
    case 'ListValue':
      return value.values.map(v => parseArgumentValue(v));
    case 'ObjectValue':
      return value.fields.reduce((obj, v) => ({
        ...obj,
        [v.name.value]: parseArgumentValue(v.value),
      }), {});
    default:
      return undefined;
  }
}

function getArgumentValue(node: FieldNode, argName: string) {
  const argument = node.arguments?.find(a => a.name.value === argName)?.value;
  return argument ? parseArgumentValue(argument) : argument;
}

function getMemberType(metaConfig: any, cubeName: string, memberName: string) {
  const cubeConfig = metaConfig.find(cube => cube.config.name === cubeName);
  if (!cubeConfig) return undefined;

  return ['measure', 'dimension'].find((memberType) => (cubeConfig.config[`${memberType}s`]
    .findIndex(entry => entry.name === `${cubeName}.${memberName}`) !== -1
  ));
}

function whereArgToQueryFilters(
  whereArg: Record<string, any>,
  prefix?: string
) {
  const queryFilters: any[] = [];

  Object.keys(whereArg).forEach((key) => {
    if (['OR', 'AND'].includes(key)) {
      queryFilters.push({
        [key.toLowerCase()]: whereArg[key].reduce(
          (filters, whereBooleanArg) => [
            ...filters,
            ...whereArgToQueryFilters(whereBooleanArg, prefix),
          ],
          []
        ),
      });
    } else if (whereArg[key].OR || whereArg[key].AND) {
      // users: {
      //   OR: {
      //     name: { equals: "Alex" }
      //     country: { equals: "US" }
      //   } # <-- a single boolean filter can be passed in directly
      //   age: { equals: 28 } # <-- will require AND
      // }
      if (Object.keys(whereArg[key]).length > 1) {
        queryFilters.push(
          ...whereArgToQueryFilters(
            {
              AND: Object.entries(whereArg[key]).reduce<any>(
                (memo, [k, v]) => [...memo, { [k]: v }],
                []
              ),
            },
            capitalize(key)
          )
        );
      } else {
        const res = whereArgToQueryFilters(whereArg[key], capitalize(key));

        queryFilters.push(...res);
      }
    } else if (prefix) {
      // handle a subfilter
      // { country: { in: ["US"] }
      Object.entries(whereArg[key]).forEach(([operator, value]) => {
        queryFilters.push({
          member: `${prefix}.${key}`,
          operator: mapWhereOperator(operator, value),
          ...(mapWhereValue(operator, value) && {
            values: mapWhereValue(operator, value),
          }),
        });
      });
    } else {
      Object.entries<any>(whereArg[key]).forEach(([member, filters]) => {
        Object.entries(filters).forEach(([operator, value]) => {
          queryFilters.push({
            member: prefix
              ? `${prefix}.${key}`
              : `${capitalize(key)}.${member}`,
            operator: mapWhereOperator(operator, value),
            ...(mapWhereValue(operator, value) && {
              values: mapWhereValue(operator, value),
            }),
          });
        });
      });
    }
  });

  return queryFilters;
}

function parseDates(result: any) {
  const { timezone } = result.query;

  const dateKeys = Object.entries<any>({
    ...result.annotation.measures,
    ...result.annotation.dimensions,
    ...result.annotation.timeDimensions,
  }).reduce((res, [key, value]) => (value.type === 'time' ? [...res, key] : res), [] as any);

  result.data.forEach(row => {
    Object.keys(row).forEach(key => {
      if (dateKeys.includes(key)) {
        row[key] = moment.tz(row[key], timezone).toISOString();
      }
      return row;
    });
  });
}

export function makeSchema(metaConfig: any) {
  const types: any[] = [
    DateTimeScalar,
    FloatFilter,
    StringFilter,
    DateTimeFilter,
    OrderBy,
    TimeDimension
  ];

  metaConfig.forEach(cube => {
    types.push(objectType({
      name: `${cube.config.name}Members`,
      definition(t) {
        cube.config.measures.forEach(measure => {
          if (measure.isVisible) {
            t.nonNull.field(safeName(measure.name), {
              type: mapType(measure.type),
              description: measure.description
            });
          }
        });
        cube.config.dimensions.forEach(dimension => {
          if (dimension.isVisible) {
            t.nonNull.field(safeName(dimension.name), {
              type: mapType(dimension.type),
              description: dimension.description
            });
          }
        });
      }
    }));

    types.push(inputObjectType({
      name: `${cube.config.name}WhereInput`,
      definition(t) {
        t.field('AND', { type: list(nonNull(`${cube.config.name}WhereInput`)) });
        t.field('OR', { type: list(nonNull(`${cube.config.name}WhereInput`)) });
        cube.config.measures.forEach(measure => {
          if (measure.isVisible) {
            t.field(safeName(measure.name), {
              type: `${mapType(measure.type, true)}Filter`,
            });
          }
        });
        cube.config.dimensions.forEach(dimension => {
          if (dimension.isVisible) {
            t.field(safeName(dimension.name), {
              type: `${mapType(dimension.type, true)}Filter`,
            });
          }
        });
      }
    }));

    types.push(inputObjectType({
      name: `${cube.config.name}OrderByInput`,
      definition(t) {
        cube.config.measures.forEach(measure => {
          if (measure.isVisible) {
            t.field(safeName(measure.name), {
              type: 'OrderBy',
            });
          }
        });
        cube.config.dimensions.forEach(dimension => {
          if (dimension.isVisible) {
            t.field(safeName(dimension.name), {
              type: 'OrderBy',
            });
          }
        });
      }
    }));
  });

  types.push(inputObjectType({
    name: 'RootWhereInput',
    definition(t) {
      t.field('AND', { type: list(nonNull('RootWhereInput')) });
      t.field('OR', { type: list(nonNull('RootWhereInput')) });
      metaConfig.forEach(cube => {
        t.field(unCapitalize(cube.config.name), {
          type: `${cube.config.name}WhereInput`
        });
      });
    }
  }));
  
  types.push(inputObjectType({
    name: 'RootOrderByInput',
    definition(t) {
      metaConfig.forEach(cube => {
        t.field(unCapitalize(cube.config.name), {
          type: `${cube.config.name}OrderByInput`
        });
      });
    }
  }));

  types.push(objectType({
    name: 'Result',
    definition(t) {
      metaConfig.forEach(cube => {
        t.nonNull.field(unCapitalize(cube.config.name), {
          type: `${cube.config.name}Members`,
          args: {
            where: arg({
              type: `${cube.config.name}WhereInput`
            }),
            orderBy: arg({
              type: `${cube.config.name}OrderByInput`
            }),
          }
        });
      });
    }
  }));

  types.push(extendType({
    type: 'Query',
    definition(t) {
      t.nonNull.field('cube', {
        type: list(nonNull('Result')),
        args: {
          where: arg({
            type: 'RootWhereInput'
          }),
          limit: intArg(),
          offset: intArg(),
          timezone: stringArg(),
          renewQuery: booleanArg(),
          orderBy: arg({
            type: 'RootOrderByInput'
          }),
        },
        resolve: async (_, { where, limit, offset, timezone, orderBy, renewQuery }, { req, apiGateway }, infos) => {
          const measures: string[] = [];
          const dimensions: string[] = [];
          const timeDimensions: any[] = [];
          let filters: any[] = [];
          const order: [string, 'asc' | 'desc'][] = [];
          
          if (where) {
            filters = whereArgToQueryFilters(where);
          }
          
          if (orderBy) {
            Object.entries<any>(orderBy).forEach(([cubeName, members]) => {
              Object.entries<any>(members).forEach(([member, value]) => {
                order.push([`${capitalize(cubeName)}.${member}`, value]);
              });
            });
          }

          getFieldNodeChildren(infos.fieldNodes[0], infos).forEach(cubeNode => {
            const cubeName = capitalize(cubeNode.name.value);
            const orderByArg = getArgumentValue(cubeNode, 'orderBy');
            // todo: throw if both RootOrderByInput and [Cube]OrderByInput provided
            if (orderByArg) {
              Object.keys(orderByArg).forEach(key => {
                order.push([`${cubeName}.${key}`, orderByArg[key]]);
              });
            }

            const whereArg = getArgumentValue(cubeNode, 'where');
            if (whereArg) {
              filters = whereArgToQueryFilters(whereArg, cubeName).concat(filters);
            }

            getFieldNodeChildren(cubeNode, infos).forEach(memberNode => {
              const memberName = memberNode.name.value;
              const memberType = getMemberType(metaConfig, cubeName, memberName);

              if (memberType === 'measure') {
                measures.push(`${cubeName}.${memberName}`);
              } else if (memberType === 'dimension') {
                const granularityNodes = getFieldNodeChildren(memberNode, infos);
                if (granularityNodes.length > 0) {
                  granularityNodes.forEach(granularityNode => {
                    const granularityName = granularityNode.name.value;
                    if (granularityName === 'value') {
                      dimensions.push(`${cubeName}.${memberName}`);
                    } else {
                      timeDimensions.push({
                        dimension: `${cubeName}.${memberName}`,
                        granularity: granularityName
                      });
                    }
                  });
                } else {
                  dimensions.push(`${cubeName}.${memberName}`);
                }
              }
            });
          });

          const query = {
            ...(measures.length && { measures }),
            ...(dimensions.length && { dimensions }),
            ...(timeDimensions.length && { timeDimensions }),
            ...(Object.keys(order).length && { order }),
            ...(limit && { limit }),
            ...(offset && { offset }),
            ...(timezone && { timezone }),
            ...(filters.length && { filters }),
            ...(renewQuery && { renewQuery }),
          };
          
          // eslint-disable-next-line no-async-promise-executor
          const results = await (new Promise<any>(async (resolve, reject) => {
            try {
              await apiGateway.load({
                query,
                queryType: QUERY_TYPE.REGULAR_QUERY,
                context: req.context,
                res: (message) => {
                  if (message.error) {
                    reject(new Error(message.error));
                  }
                  resolve(message);
                },
              });
            } catch (e) {
              reject(e);
            }
          }));

          parseDates(results);

          return results.data.map(entry => R.toPairs(entry)
            .reduce((res, pair) => {
              let path = pair[0].split('.');
              path[0] = unCapitalize(path[0]);
              if (results.annotation.dimensions[pair[0]]?.type === 'time') {
                path = [...path, 'value'];
              }
              return (results.annotation.timeDimensions[pair[0]] && path.length !== 3)
                ? res : R.set(R.lensPath(path), pair[1], res);
            }, {}));
        }
      });
    }
  }));

  return nexusMakeSchema({ types });
}
