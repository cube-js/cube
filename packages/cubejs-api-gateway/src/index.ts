export * from './gateway';
export * from './interfaces';
export * from './CubejsHandlerError';
export * from './UserError';
export { getRequestIdFromRequest } from './requestParser';
