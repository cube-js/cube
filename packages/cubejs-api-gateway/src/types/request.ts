/**
 * @license Apache-2.0
 * @copyright Cube Dev, Inc.
 * @fileoverview
 * Network request data types definition.
 */

import type { Request as ExpressRequest } from 'express';
import { RequestType, ApiType, ResultType } from './strings';
import { Query } from './query';

/**
 * Network request context interface.
 * @todo requestId should be described in strings.
 * @todo securityContext description.
 * @todo why this is interface?
 */
interface RequestContext {
  securityContext: any;
  requestId: string;
  signedWithPlaygroundAuthSecret?: boolean;
  appName?: string,
  protocol?: string,
  apiType?: string,
}

/**
 * Additional request data record.
 * @todo any could be changed to unknown?
 * @todo Maybe we can add type limitations?
 */
type RequestExtension = Record<string, any>;

/**
 * Network request completed context data type.
 */
type ExtendedRequestContext =
  RequestContext & RequestExtension;

/**
 * Gateway request interface.
 */
interface Request extends ExpressRequest {
  context?: ExtendedRequestContext,
  signedWithPlaygroundAuthSecret?: boolean,

  /**
   * Security context object. Should be used instead of deprecated
   * Request#authInfo.
   * @todo any could be changed to unknown?
   * @todo Maybe we can add type limitations?
   */
  securityContext?: any,

  /**
   * @deprecated
   */
  authInfo?: any,
}

/**
 * Function that should provides basic query conversion mechanic.
 * Used as a part of a main configuration object of the server-core
 * to provide extendabillity to a query processing logic.
 */
type QueryRewriteFn =
  (query: Query, context: RequestContext) => Promise<Query>;

/**
 * Function that should provides a logic for extracting security
 * context from the request. Used as a part of a main configuration
 * object of the server-core to provide extendabillity to a query
 * security processing logic.
 * @todo any could be changed to unknown?
 * @todo Maybe we can add type limitations?
 */
type SecurityContextExtractorFn =
  (ctx: Readonly<RequestContext>) => any;

/**
 * Function that should provides a logic for extracting request
 * extesion context from the request. Used as a part of a main
 * configuration object of the server-core to provide extendabillity
 * to a query processing logic.
 */
type ExtendContextFn =
  (req: ExpressRequest) =>
    Promise<RequestExtension> | RequestExtension;

/**
 * Function that should provides a logic for the response result
 * processing. Used as a part of a main configuration object of the
 * server-core to provide extendabillity for this logic.
 * @todo any could be changed to unknown?
 * @todo Maybe we can add type limitations?
 */
type ResponseResultFn =
  (
    message: Record<string, any> | Record<string, any>[],
    extra?: { status: number }
  ) => void;

/**
 * Base HTTP request parameters map data type.
 * @todo map it to Request.
 */
type BaseRequest = {
  context: RequestContext;
  res: ResponseResultFn
};

/**
 * Data query HTTP request parameters map data type.
 */
type QueryRequest = BaseRequest & {
  query: Record<string, any> | Record<string, any>[];
  queryType?: RequestType;
  apiType?: ApiType;
  resType?: ResultType
};

/**
 * Pre-aggregations selector object.
 */
type PreAggsSelector = {
  contexts?: {securityContext: any}[],
  timezones: string[],
  dataSources?: string[],
  cubes?: string[],
  preAggregations?: string[],
};

/**
 * Posted pre-aggs job object.
 */
type PreAggJob = {
  request: string;
  context: {securityContext: any};
  preagg: string;
  table: string;
  target: string;
  structure: string;
  content: string;
  updated: number;
  key: any[];
  status: string;
  timezone: string;
  dataSource: string;
};

/**
 * The `/cubejs-system/v1/pre-aggregations/jobs` endpoint object type.
 */
type PreAggsJobsRequest = {
  action: 'post' | 'get' | 'delete',
  selector?: PreAggsSelector,
  tokens?: string[]
  resType?: 'object' | 'array'
};

type PreAggJobStatusItem = {
  token: string;
  table: string;
  status: string;
  selector: PreAggsSelector;
};

type PreAggJobStatusObject = {
  [token: string]: {
    table: string;
    status: string;
    selector: PreAggsSelector;
  }
};

type PreAggJobStatusResponse =
  | PreAggJobStatusItem[]
  | PreAggJobStatusObject;

export {
  RequestContext,
  RequestExtension,
  ExtendedRequestContext,
  Request,
  QueryRewriteFn,
  SecurityContextExtractorFn,
  ExtendContextFn,
  ResponseResultFn,
  BaseRequest,
  QueryRequest,
  PreAggsJobsRequest,
  PreAggsSelector,
  PreAggJob,
  PreAggJobStatusItem,
  PreAggJobStatusObject,
  PreAggJobStatusResponse,
};
