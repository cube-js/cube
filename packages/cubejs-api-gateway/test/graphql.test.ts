/* globals describe,test,expect */

import bodyParser from 'body-parser';
// eslint-disable-next-line import/no-extraneous-dependencies
import express from 'express';
import { graphqlHTTP } from 'express-graphql';
import { GraphQLObjectType } from 'graphql';
// eslint-disable-next-line import/no-extraneous-dependencies
import fs from 'fs-extra';
// eslint-disable-next-line import/no-extraneous-dependencies
import request from 'supertest';

import { makeSchema } from '../src/graphql';

const metaConfig = [
  {
    config: {
      name: 'Orders',
      measures: [
        {
          name: 'Orders.count',
          isVisible: true,
        },
      ],
      dimensions: [
        {
          name: 'Orders.id',
          isVisible: true,
        },
        {
          name: 'Orders.status',
          type: 'string',
          isVisible: true,
        },
        {
          name: 'Orders.createdAt',
          type: 'time',
          isVisible: true,
        },
      ],
    },
  },
  {
    config: {
      name: 'Users',
      measures: [
        {
          name: 'Users.count',
          isVisible: true,
        },
      ],
      dimensions: [
        {
          name: 'Users.id',
          isVisible: true,
        },
        {
          name: 'Users.city',
          type: 'string',
          isVisible: true,
        },
        {
          name: 'Users.createdAt',
          type: 'time',
          isVisible: true,
        },
      ],
    },
  },
];

function gqlQuery(query: string, variables?: Record<string, string | number>) {
  return JSON.stringify({
    operationName: 'CubeQuery',
    query,
    variables,
  });
}

function expectValidSchema(schema) {
  expect(schema).toBeDefined();
  expect(schema.getTypeMap()).toHaveProperty('OrdersMembers');
  const ordersFields = (schema.getType('OrdersMembers') as GraphQLObjectType).getFields();
  expect(ordersFields).toHaveProperty('id');
  expect(ordersFields).toHaveProperty('status');
  expect(ordersFields).toHaveProperty('createdAt');
}

const app = express();

const jsonParser = bodyParser.json({ limit: '1mb' });

app.use('/graphql', jsonParser, (req, res) => {
  const schema = makeSchema(metaConfig);

  return graphqlHTTP({
    schema,
    context: {
      req,
      apiGateway: {
        async load({ query, res: response }) {
          expect(query).toMatchSnapshot(req.body.query);

          response({
            query,
            annotation: {},
            data: [],
          });
        },
      },
    },
  })(req, res);
});

const GRAPHQL_QUERIES_PATH = `${process.cwd()}/test/graphql-queries/base.gql`;

describe('GraphQL Schema', () => {
  test('should make valid schema', () => {
    const schema = makeSchema(metaConfig);
    expectValidSchema(schema);
  });

  test('should make valid schema when name is not capitalized', async () => {
    const schema = makeSchema(JSON.parse(
      JSON.stringify(metaConfig)
        .replace(/Orders/g, 'orders')
    ));
    expectValidSchema(schema);
  });
  
  const queries = fs.readFileSync(GRAPHQL_QUERIES_PATH, 'utf-8').split('\n\n');

  queries.forEach((query, index) => {
    test(`GraphQL query ${index}`, async () => {
      const { error } = await request(app)
        .post('/graphql')
        .set('Content-Type', 'application/json')
        .send(gqlQuery(query));

      expect((<any>error)?.text).toBeUndefined();
    });
  });
});
