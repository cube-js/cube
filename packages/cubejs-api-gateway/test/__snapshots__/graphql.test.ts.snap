// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphQL Schema GraphQL query 0: query CubeQuery {
  cube {
    orders {
      count
      createdAt {
        day
      }
    }
  }
} 1`] = `
Object {
  "measures": Array [
    "Orders.count",
  ],
  "timeDimensions": Array [
    Object {
      "dimension": "Orders.createdAt",
      "granularity": "day",
    },
  ],
}
`;

exports[`GraphQL Schema GraphQL query 1: query CubeQuery {
  cube(where: { orders: { status: { equals: "shipped" } } }) {
    orders {
      count
      createdAt {
        day
      }
    }
  }
} 1`] = `
Object {
  "filters": Array [
    Object {
      "member": "Orders.status",
      "operator": "equals",
      "values": Array [
        "shipped",
      ],
    },
  ],
  "measures": Array [
    "Orders.count",
  ],
  "timeDimensions": Array [
    Object {
      "dimension": "Orders.createdAt",
      "granularity": "day",
    },
  ],
}
`;

exports[`GraphQL Schema GraphQL query 2: query CubeQuery {
  cube(
    where: {
      orders: { createdAt: { inDateRange: ["2022-01-01", "2022-02-01"] } }
    }
  ) {
    orders(orderBy: { createdAt: asc }) {
      count
      createdAt {
        value
      }
    }
  }
} 1`] = `
Object {
  "dimensions": Array [
    "Orders.createdAt",
  ],
  "filters": Array [
    Object {
      "member": "Orders.createdAt",
      "operator": "inDateRange",
      "values": Array [
        "2022-01-01",
        "2022-02-01",
      ],
    },
  ],
  "measures": Array [
    "Orders.count",
  ],
  "order": Array [
    Array [
      "Orders.createdAt",
      "asc",
    ],
  ],
}
`;

exports[`GraphQL Schema GraphQL query 3: query CubeQuery {
  cube(
    where: {
      orders: { createdAt: { inDateRange: ["2022-01-01", "2022-02-01"] } }
    }
  ) {
    orders(orderBy: { createdAt: asc }) {
      count
      createdAt {
        day
      }
    }
  }
} 1`] = `
Object {
  "filters": Array [
    Object {
      "member": "Orders.createdAt",
      "operator": "inDateRange",
      "values": Array [
        "2022-01-01",
        "2022-02-01",
      ],
    },
  ],
  "measures": Array [
    "Orders.count",
  ],
  "order": Array [
    Array [
      "Orders.createdAt",
      "asc",
    ],
  ],
  "timeDimensions": Array [
    Object {
      "dimension": "Orders.createdAt",
      "granularity": "day",
    },
  ],
}
`;

exports[`GraphQL Schema GraphQL query 4: query CubeQuery {
  cube(where: { orders: { createdAt: { inDateRange: "This month" } } }) {
    orders {
      count
      createdAt {
        value
      }
    }
  }
} 1`] = `
Object {
  "dimensions": Array [
    "Orders.createdAt",
  ],
  "measures": Array [
    "Orders.count",
  ],
  "timeDimensions": Array [
    Object {
      "dateRange": "This month",
      "dimension": "Orders.createdAt",
    },
  ],
}
`;

exports[`GraphQL Schema GraphQL query 5: query CubeQuery {
  cube(where: { orders: { createdAt: { inDateRange: "This month" } } }) {
    orders {
      count
      createdAt {
        day
      }
    }
  }
} 1`] = `
Object {
  "measures": Array [
    "Orders.count",
  ],
  "timeDimensions": Array [
    Object {
      "dateRange": "This month",
      "dimension": "Orders.createdAt",
      "granularity": "day",
    },
  ],
}
`;

exports[`GraphQL Schema GraphQL query 6: query CubeQuery {
  cube(where: { orders: { createdAt: { inDateRange: "2 weeks ago to now" } } }) {
    orders {
      count
      createdAt {
        day
      }
    }
  }
} 1`] = `
Object {
  "measures": Array [
    "Orders.count",
  ],
  "timeDimensions": Array [
    Object {
      "dateRange": "2 weeks ago to now",
      "dimension": "Orders.createdAt",
      "granularity": "day",
    },
  ],
}
`;

exports[`GraphQL Schema GraphQL query 7: query CubeQuery {
  cube(
    where: {
      orders: { createdAt: { notInDateRange: ["2022-01-01", "2022-02-01"] } }
    }
  ) {
    orders {
      count
      createdAt {
        day
      }
    }
  }
} 1`] = `
Object {
  "filters": Array [
    Object {
      "member": "Orders.createdAt",
      "operator": "notInDateRange",
      "values": Array [
        "2022-01-01",
        "2022-02-01",
      ],
    },
  ],
  "measures": Array [
    "Orders.count",
  ],
  "timeDimensions": Array [
    Object {
      "dimension": "Orders.createdAt",
      "granularity": "day",
    },
  ],
}
`;

exports[`GraphQL Schema GraphQL query 8: query CubeQuery {
  cube {
    orders(orderBy: { count: desc, status: asc }) {
      count
      status
      createdAt {
        value
      }
    }
    users(orderBy: { city: desc }) {
      city
    }
  }
} 1`] = `
Object {
  "dimensions": Array [
    "Orders.status",
    "Orders.createdAt",
    "Users.city",
  ],
  "measures": Array [
    "Orders.count",
  ],
  "order": Array [
    Array [
      "Orders.count",
      "desc",
    ],
    Array [
      "Orders.status",
      "asc",
    ],
    Array [
      "Users.city",
      "desc",
    ],
  ],
}
`;

exports[`GraphQL Schema GraphQL query 9: query CubeQuery {
  cube(
    where: {
      orders: { status: { in: ["canceled", "active"] } }
      OR: [
        {
          users: { OR: [{ city: { set: false } }, { city: { equals: "US" } }] }
        }
      ]
    }
  ) {
    orders {
      count
      status
    }
  }
}
 1`] = `
Object {
  "dimensions": Array [
    "Orders.status",
  ],
  "filters": Array [
    Object {
      "or": Array [
        Object {
          "or": Array [
            Object {
              "member": "Users.city",
              "operator": "notSet",
            },
            Object {
              "member": "Users.city",
              "operator": "equals",
              "values": Array [
                "US",
              ],
            },
          ],
        },
      ],
    },
    Object {
      "member": "Orders.status",
      "operator": "equals",
      "values": Array [
        "canceled",
        "active",
      ],
    },
  ],
  "measures": Array [
    "Orders.count",
  ],
}
`;
